run/websockets/redis.js

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const {promisify} = require('util');
+
+// [START cloudrun_websockets_redis]
+const REDISHOST = process.env.REDISHOST || 'localhost';
+const REDISPORT = process.env.REDISPORT || 6379;
+
+// Instantiate the Redis client
+const redisClient = require('redis').createClient(REDISPORT, REDISHOST);
+// [END cloudrun_websockets_redis]
+
+// Set up to use promises and async/await for Redis methods
+const redisGet = promisify(redisClient.get).bind(redisClient);
+const redisExists = promisify(redisClient.exists).bind(redisClient);
+
+// Insert new messages into the Redis cache
+async function addMessageToCache(roomName, msg) {
+  let room = await getRoomFromCache(roomName);
+  if (room) {
+    // Update old room
+    room.messages.push(msg);
+  } else {
+    // Create a new room
+    room = {
+      room: roomName,
+      messages: [msg],
+    };
+  }
+  redisClient.set(roomName, JSON.stringify(room));
+  addMessageToDb(room);
+}
+
+// Insert messages into the database
+async function addMessageToDb(data) {
+  const room = messageDb.find(messages => messages.room === data.room);
+  if (room) {
+    // Update room in database
+    Object.assign(room, data);
+  } else {
+    // Create new room in database
+    messageDb.push(data);
+  }
+}
+
+// Query Redis for messages for a specific room
+// If not in Redis, query the database
+async function getRoomFromCache(roomName) {
+  if (!(await redisExists(roomName))) {
+    const room = getRoomFromDatabase(roomName);
+    if (room) {
+      redisClient.set(roomName, JSON.stringify(room));
+    }
+  }
+  return JSON.parse(await redisGet(roomName));
+}
+
+// Query the database for messages for a specific room
+function getRoomFromDatabase(roomName) {
+  return messageDb.find(messages => messages.room === roomName);
+}
+
+// Example of local database storage -
+// Production applications should use a persistent database such as Firestore
```

Comment:
suggestion: Something I find a bit confusing about this, is I'm used to regarding in-memory data when Redis is in use as a "local cache" in front of redis. In this case, it seems to be that this is the canonical data and redis is being used to synchronize in-memory data between instances.

I think this needs some more words to explain, and even if it's covered in the tutorial I think more comments are needed here to make sure maintenance needs are clear. It might also be helpful to separate out the "database" code from the "cache" code.

Reply:
I am trying to mock having persistent storage such as SQL or firestore, so I don't have to add another dependency. However, I see that this can be confusing since it's in-memory therefore not really persistent. I will try updating the comments to make this clear.